Source/engine.cpp

```diff
@@ -217,72 +193,64 @@ void CelBlitLightSafeTo(CelOutputBuffer out, int sx, int sy, BYTE *pRLEBytes, in
 			} else {
 				width = -(char)width;
 				dst += width;
-				i -= width;
+				w -= width;
 			}
 		}
 	}
 }
 
 void CelBlitLightTransSafeTo(CelOutputBuffer out, int sx, int sy, BYTE *pRLEBytes, int nDataSize, int nWidth)
 {
-	int w;
-	bool shift;
-	BYTE *tbl;
-
 	assert(pRLEBytes != nullptr);
-	int i;
-	BYTE width;
-	BYTE *src, *dst;
-
-	src = pRLEBytes;
-	dst = out.at(sx, sy);
-	tbl = &pLightTbl[light_table_index * 256];
-	w = nWidth;
-	shift = (BYTE)(size_t)dst & 1;
-
-	for (; src != &pRLEBytes[nDataSize]; dst -= out.pitch() + w, shift = (shift + 1) & 1) {
-		for (i = w; i;) {
-			width = *src++;
-			if (!(width & 0x80)) {
-				i -= width;
+
+	BYTE *src = pRLEBytes;
+	BYTE *dst = out.at(sx, sy);
+	BYTE *tbl = &pLightTbl[light_table_index * 256];
+	BYTE shift = ((BYTE)(size_t)dst & 1) != 0;

```

Comment:
Took me a bit to grok that.

This checks if the destination pointer is pointing to an even address, i.e. 16 bit aligned.

To do so, it first casts the pointer to size_t, then immediately to BYTE, then it masks out the lowest bit, and if the result is not 0 (so 1), then the expression returns true - which is immediately converted into a BYTE again, for some later use.

```suggestion
	BYTE shift = ((BYTE)(size_t)dst & 1) != 0;
	bool shift = (static_cast<uintptr_t>(dst) % 2) != 0;
```

Reply:
`error: invalid ‘static_cast’ from type ‘devilution::BYTE*’ {aka ‘unsigned char*’}`

Reply:
Did a c-style cast for now `(size_t)dst`, but applied your other suggestions. Then we can make it in to a proper C++ style cast later :)

Reply:
Woops, reinterpret_cast

Reply:
Yeah, now I don't feel bad for not understanding when to use them either...